### PR TITLE
Client: It is best to judge whether the layout is valid or not before using in->layout.

### PR DESCRIPTION
**Pull request by liwenpeng19**
…n->layout.

Signed-off-by: wenpengLi  liwenpeng@inspur.com

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -872,7 +872,11 @@ Inode * Client::add_update_inode(InodeStat *st, utime_t from,
 
   if (new_version ||
       (new_issued & (CEPH_CAP_ANY_FILE_RD | CEPH_CAP_ANY_FILE_WR))) {
-    in->layout = st->layout;
+    if(st->layout.is_valid()){
+      in->layout = st->layout;
```
Comment (batrick) [inline src/client/Client.cc]: What circumstance would the `InodeStat::layout` from the MDS not be valid?

File: src/client/Client.cc [hunk 2]
```diff
@@ -872,7 +872,11 @@ Inode * Client::add_update_inode(InodeStat *st, utime_t from,
...
+    } else {
+      ldout(cct, 10) << " bad layout " << dendl;
```
Comment (batrick) [inline src/client/Client.cc]: and leave it unmodified??

File: src/client/Client.cc [hunk 3]
```diff
@@ -5206,9 +5210,13 @@ void Client::handle_cap_grant(MetaSession *session, Inode *in, Cap *cap, const M
   }
 
   if (new_caps & (CEPH_CAP_ANY_FILE_RD | CEPH_CAP_ANY_FILE_WR)) {
-    in->layout = m->get_layout();
-    update_inode_file_size(in, issued, m->get_size(),
-			   m->get_truncate_seq(), m->get_truncate_size());
+    if((m->get_layout()).is_valid()){
```
Comment (batrick) [inline src/client/Client.cc]: same question

File: src/client/Client.cc (hunk 1 above)
Comment (liwenpeng19) [reply src/client/Client.cc]: We've seen core files generated in real projects because layout is invalid.

File: src/client/Client.cc (hunk 2 above)
Comment (liwenpeng19) [reply src/client/Client.cc]: It is best to judge whether the layout is valid.

File: src/client/Client.cc (hunk 1 above)
Comment (batrick) [reply src/client/Client.cc]: The layout should never be invalid and the client should not be "falling back" to some default layout. Any fix for this should be done in the MDS.

Also, to be clear, you haven't answered my question. Can you give an example situation where the file layout would be invalid.

Comment (liwenpeng19) [reply src/client/Client.cc]: Repeat write and truncate 0 for a file ,there is a chance  to appear

**Pull request closed without merging**